### DISCUSSION TITLE
Ability to exclude files from those matched by the package.json:files field

### DISCUSSION THREAD

**Discussion by trusktr** [npm/feedback#195] [CLI] [answered]
Trying publish only one folder, while excluding only certain files from that folder, is very difficult with `npm`.

Say I want to include only `dist/`, but `dist/` also happens to contain `.test.js` files, and `dist/` is an auto-generated folder.

If we place

```
{
  "files": ["dist/"]
}
```

in `package.json`, there is no available way to ignore the any `.test.js` files inside of `dist/`.

If we add a `.npmignore` file, like

```conf
/**/*.test.js
```

it will not ignore the test files because they are already included by `files` in package.json. `npm pack --dry-run` will lit those files as being included.

An `excludeFiles` field for `package.json` would be very convenient. It would allows us to filter from included files. This is a common sort of thing in many other tools, where it is possible to easily exclude files from the set of already-included files. 

From my experience, this is usually this is what people want and is the most intuitive. Webpack is a good example of a widely-used tool that is configurable in that manner, while `npm` does things the opposite way only (prefers deny-listing first instead of allow-listing first).

---

Furthermore, I tried deleting the `files` field and using only this `.npmignore` file:

```conf
# ignore everything first
/**/*

# and include only this stuff
!/dist/
!/src/

# but filter out this stuff:
/**/*.test.*
```

and that fails miserably. The point is, it is hard to achieve what is desired: to include a set of files, then to exclude from the set of included files.

---

Did I miss something? An `excludeFiles` field that makes things easy like with other tools would go a long way.

**Comment by trusktr** [comment]
What finally worked was using only `.npmignore` with rules in order of specificity:

```conf
/**/*

# these two rules are more specific than the last.
!/dist/**/*
!/src/**/*

# these two rules are more specific than the last.
/dist/**/*.test.*
/src/**/*.test.*
```

But this is not as intuitive intuitive as `excludeFiles` would be. An `excludeFiles` field would make things easy like with other tools: anything matched by `excludeFiles` simply excludes from what is already matched by `files`, with no specificity needed.

It would look like this:

```json
{
  "files": ["dist/", "src/"],
  "excludeFiles": ["**/*.test.*"],
}
```

and it would simply work. YOu can see it is shorter than the `.npmignore` version, and it will remain a lot shorter with more complex cases that will cause the current `.npmignore` solution to grow much bigger.

In that example, first anything in `dist` and `src` would be included. Then any files that have `.test.` in them would be discluded based on the `excludeFiles` pattern. Simple.

**Comment by MylesBorins** [comment, marked as answer]
hey @trusktr if you still want to pursue this work you should open an issue over at https://github.com/npm/rfcs/issues
